pkg/icingadb/history/retention.go

```diff
@@ -0,0 +1,133 @@
+package history
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icingadb/pkg/icingadb"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/periodic"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"time"
+)
+
+// RetentionOptions defines the non-default mapping of history tables with their retention period in days.
+type RetentionOptions map[string]uint
+
+// Validate checks constraints in the supplied retention options and
+// returns an error if they are violated.
+func (o RetentionOptions) Validate() error {
+	for table := range o {
+		if _, ok := statements[table]; !ok {
+			return errors.Errorf("invalid table %s for history retention", table)
+		}
+	}
+
+	return nil
+}
+
+// Retention deletes rows from history tables that exceed their configured retention period.
+type Retention struct {
+	db          *icingadb.DB
+	logger      *logging.Logger
+	days        uint
+	interval    time.Duration
+	count       uint
+	gracePeriod time.Duration
+	options     RetentionOptions
+}
+
+// NewRetention returns a new Retention.
+func NewRetention(db *icingadb.DB, days uint, interval time.Duration, count uint, gracePeriod time.Duration, options RetentionOptions, logger *logging.Logger) *Retention {
+	return &Retention{
+		db:          db,
+		logger:      logger,
+		days:        days,
+		interval:    interval,
+		count:       count,
+		gracePeriod: gracePeriod,
+		options:     options,
+	}
+}
+
+// Start starts the retention.
+func (r *Retention) Start(ctx context.Context) error {
+	tasks := make([]periodic.Stopper, 0, len(statements))
+	defer func() {
+		for i := range tasks {
+			tasks[i].Stop()
+		}
+	}()
```

Comment:
Could the same also be achieved by cancelling a context passed to `periodic`?

Reply:
Yes, but I would have to call the cancel function in a defer anyway and it shouldn't be done in the for loop. Or do I misunderstand you?

Reply:
If I'm not missing something,
```go
ctx, cancel := context.WithCancel(ctx)
defer cancel()
```
should do basically the same here (with that new `ctx` passed to `periodic.Start()` below) as is will stop all tasks with one cancel operation. Also looks like you won't even need the `tasks` variable anymore to keep track of them then.

Reply:
Nice, that looks good. I will check it.